src/noit_message_decoder.c

```diff
@@ -461,7 +462,10 @@ noit_metric_tags_parse_one(const char *tagnm, size_t tagnmlen,
 static int
 tag_canonical_size(noit_metric_tag_t *tag) {
   int len;
-  char dbuff[NOIT_TAG_MAX_PAIR_LEN];
+  // TODO: array increased from 256 to 4096+7 to allow for longer tag names.
```

Comment:
Memory wise it does not matter because it's on the stack and that's pretty much free.  However, nowhere does this check if ``max_tag_length`` is an acceptable value.  The use of ``mtev_dyn_buffer_t`` would be appropriate in this circumstances

Reply:
> The use of `mtev_dyn_buffer_t` would be appropriate in this circumstances

I will replace
```
char dbuff[NOIT_TAG_MAX_PAIR_LEN];
```
with `mtev_dyn_buffer_t`

> nowhere does this check if `max_tag_length` is an acceptable value. 

assuming this comment refers to `noit_metric_tags_parse_one`, would it be safer to implement the `max_tag_length` parameter as an enum so that it can only be 256 or 4096?

Reply:
No because someone else might need a value that's different than that and this is a private function

Reply:
> nowhere does this check if `max_tag_length` is an acceptable value.

if a `max_tag_length` > 4096 is passed to the function, should we return as a failure or manually reduce the `max_tag_length` to 4096?

Reply:
That doesn't matter, what matters is the array being large enough

```c
char dbuff[NOIT_TAG_MAX_PAIR_LEN];
```
in combination with a value higher than 4k would corrupt the stack but you're using ``mtev_dyn_buffer`` now

Reply:
Ah, I understand. Thanks!

Reply:
@pjulien I can not pass a `mtev_dyn_buffer_t` to `noit_metric_tagset_decode_tag` because they would mutate the buffer without ensuring that the buffer is big enough to handle the new data.
using `mtev_dyn_buffer_t` would require writing a new function to handle the functionality of `noit_metric_tagset_decode_tag` for a `mtev_dyn_buffer_t`. The same would need to be done for all functions that work with the other arrays in the code that are currently sized for the max tag length (256) and need to also be updated to support `__name`.
This isn't very practical, so I think we should, instead, perform a check to make sure that the requested `max_tag_length` is less than the size of the array (4096) and also increase the array sizes to 65535 as @pamaddox pointed out is already being done within `check_metadata_alter_specific`:
```
  char decoded[NOIT_TAG_MAX_PAIR_LEN + 1];
  size_t decoded_len =
    noit_metric_tagset_decode_tag(decoded, sizeof(decoded) - 1, tag, strlen(tag));
```
https://circonus.slack.com/archives/C0J5RAKDJ/p1669845638697289

Reply:
Use ``mtev_dyn_buffer_ensure``